app/PaymentDrivers/CheckoutCom/CreditCard.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+/**
+ * Invoice Ninja (https://invoiceninja.com).
+ *
+ * @link https://github.com/invoiceninja/invoiceninja source repository
+ *
+ * @copyright Copyright (c) 2020. Invoice Ninja LLC (https://invoiceninja.com)
+ *
+ * @license https://opensource.org/licenses/AAL
+ */
+
+namespace App\PaymentDrivers\CheckoutCom;
+
+use App\Http\Requests\ClientPortal\Payments\PaymentResponseRequest;
+use App\Models\PaymentHash;
+use App\PaymentDrivers\CheckoutComPaymentDriver;
+use Checkout\Models\Payments\IdSource;
+use Checkout\Models\Payments\Payment;
+use Checkout\Models\Payments\TokenSource;
+
+class CreditCard
+{
+    use Utilities;
+
+    /**
+     * @var \App\PaymentDrivers\CheckoutComPaymentDriver
+     */
+    public $checkout;
+
+    public function __construct(CheckoutComPaymentDriver $checkout)
+    {
+        $this->checkout = $checkout;
+    }
+
+    /**
+     * An authorization view for credit card.
+     * 
+     * @param mixed $data 
+     * @return \Illuminate\Contracts\View\Factory|\Illuminate\View\View 
+     */
+    public function authorizeView($data)
+    {
+        return render('gateways.checkout.credit_card.authorize');
+    }
+
+    /**
+     * Checkout.com supports doesn't support direct authorization of the credit card.
+     * Token can be saved after the first (successful) purchase.
+     * 
+     * @param mixed $data 
+     * @return void 
+     */
+    public function authorizeResponse($data)
+    {
+        return;
+    }
+
+    public function paymentView($data)
+    {
+        $data['gateway'] = $this->checkout;
+        $data['company_gateway'] = $this->checkout->company_gateway;
+        $data['client'] = $this->checkout->client;
+        $data['currency'] = $this->checkout->client->getCurrencyCode();
+        $data['value'] = $this->checkout->convertToCheckoutAmount($data['amount_with_fee'], $this->checkout->client->getCurrencyCode());
+        $data['raw_value'] = $data['amount_with_fee'];
+        $data['customer_email'] = $this->checkout->client->present()->email;
+
+        return render('gateways.checkout.credit_card.pay', $data);
+    }
+
+    public function paymentResponse(PaymentResponseRequest $request)
+    {
+        $this->checkout->init();
+
+        $state = [
+            'server_response' => json_decode($request->gateway_response),
+            'value' => $request->value,
+            'raw_value' => $request->raw_value,
+            'currency' => $request->currency,
+            'payment_hash' => $request->payment_hash,
+            'reference' => $request->payment_hash,
+        ];
+
+        $state = array_merge($state, $request->all());
+        $state['store_card'] = boolval($state['store_card']);
+
+        $payment_hash = PaymentHash::whereRaw('BINARY `hash`= ?', [$request->payment_hash])->first();
```

Comment:
Lets nest this outside in the basedriver?

Reply:
Solved here: https://github.com/invoiceninja/invoiceninja/pull/4202/commits/223ae4cb5bf1823885cf744ce84123f0e654cac2#diff-9229767762db39ad236f5c2a30a13fa853f7f72e4eb1a072ae448ea7d4067493R124-R130